### PR TITLE
[866] Improve the process for adding rake tasks in Terraform

### PR DESCRIPTION
**Pull request by erbridge**
## Trello card URL:

https://trello.com/c/aOPRDDgM/866-improve-the-process-for-adding-one-off-tasks-in-terraform

## Changes in this PR:

Refactor the ECS task Terraform. Due to how Terraform tracks state, this requires manipulating the state to avoid thrashing and make the actual changes visible. `bin/terraform-migrate` should do that state manipulation, so only real changes should appear in a plan.

## Next steps:

- [ ] Run `terraform init`
- [ ] Run `bin/terraform-migrate <workspace>`
- [ ] Deploy Terraform
- [ ] Clean out `bin/terraform-migrate` after all environments are updated

### CODE REVIEW

**Comment by erbridge** [comment]
[Improve the process for adding one-off tasks in Terraform](https://trello.com/c/aOPRDDgM/866-improve-the-process-for-adding-one-off-tasks-in-terraform)

File: terraform/modules/container-definitions/rake-container-definition/rake-container-definition.tf [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+    database_url             = "${var.database_url}"
+    elastic_search_url       = "${var.elastic_search_url}"
+    aws_elasticsearch_region = "${var.aws_elasticsearch_region}"
+    aws_elasticsearch_key    = "${var.aws_elasticsearch_key}"
+    aws_elasticsearch_secret = "${var.aws_elasticsearch_secret}"
+    feature_import_vacancies = "${var.feature_import_vacancies}"
+
+    rollbar_access_token = "${var.rollbar_access_token}"
+
+    # We are creating a partially rendered template from a template file.
```
Comment (adrianclay) [inline terraform/modules/container-definitions/rake-container-definition/rake-container-definition.tf]: Templateception.

File: bin/terraform-migrate [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# One-off ECS rake tasks
+
+terraform state mv module.ecs.aws_ecs_task_definition.reindex_vacancies_task module.ecs.module.reindex_vacancies_task.aws_ecs_task_definition.main || true
+terraform state mv module.ecs.aws_ecs_task_definition.seed_vacancies_from_api_task module.ecs.module.seed_vacancies_from_api_task.aws_ecs_task_definition.main || true
+terraform state mv module.ecs.aws_ecs_task_definition.migrate_phase_to_phases_task module.ecs.module.migrate_phase_to_phases_task.aws_ecs_task_definition.main || true
+terraform state mv module.ecs.aws_ecs_task_definition.performance_platform_submit_all_task module.ecs.module.performance_platform_submit_all_task.aws_ecs_task_definition.main || true
+
+# Scheduled ECS rake tasks
+
+terraform state mv module.ecs.aws_ecs_task_definition.sessions_trim_task module.ecs.module.sessions_trim_task.aws_ecs_task_definition.main || true
```
Comment (adrianclay) [inline bin/terraform-migrate]: How did you generate these lines was it manually, or is there a tool which can help with that?

File: terraform/modules/container-definitions/rake-container-definition/rake-container-definition.tf (hunk 1 above)
Comment (erbridge) [reply terraform/modules/container-definitions/rake-container-definition/rake-container-definition.tf]: Exactly :)

File: bin/terraform-migrate (hunk 2 above)
Comment (erbridge) [reply bin/terraform-migrate]: I created them manually as I was doing the refactoring, confirming that in each step I had a noop plan (or nearly noop in the case of changed descriptions etc.).

**Comment by erbridge** [comment]
@adrianclay I was messing with `edge` while developing this. It's worth noting that this only has a forward migration, so rolling back is a little manual.

Suggest we pair on merging this on Tuesday.

**Comment by adrianclay** [comment]
Hmm, yep lets discuss in person on Tuesday.  I _assumed_ there is some version control for the terraform state in S3.  Could look into whether that is the case, and what the process for restoring is.

**Comment by erbridge** [comment]
Oh of course, running `terraform state mv` creates local backup states you can rollback to. The problem we had was that I left the migrated state on `edge` when I was done and the next `terraform-deploy` run on that environment had some trouble running. I should have rolled back rather than leaving it.

**Comment by erbridge** [comment]
Replanned on `edge`. Looks good.

**Comment by erbridge** [comment]
Rolling back the state is only any good if you haven't actually applied. That just pushes a new state file, meaning you'd end up with a bunch of duplicates if you've deployed. Safest thing would be to do at that point would be to do the reverse of all the `mv` operations.

**Pull request merged**